### PR TITLE
Engines now return FileMeta with correct millisecond timestamps

### PR DESCRIPTION
**Pull request by OussamaSaoudi-db**
## What changes are proposed in this pull request?
This PR fixes the `list_from` method used by both the `DefaultEngine` and the `SyncEngine` to return the correct `FileMeta` struct. Previously, both the engines would return the timestamp in number of seconds from Unix Epoch. However, `FileMeta` specifies that we should expect the modification time to be in milliseconds since Unix Epoch:
```
pub struct FileMeta {
    /// The fully qualified path to the object
    pub location: Url,
    /// The last modified time as milliseconds since unix epoch
    pub last_modified: i64,
    /// The size in bytes of the object
    pub size: usize,
}
```
### This PR affects the following public APIs
This changes the behaviour of the following:
- `SyncFilesystemClient::list_from`
- `ObjectStoreFileSystemClient::list_from`
- any transitive users of these methods

## How was this change tested?
I compare the `FileMeta` returned by each of the engines to expected `FileMeta` within 10 seconds of creation time.

### CODE REVIEW

**Comment by OussamaSaoudi-db** [comment]
Since I'm using the filesystem to get a ground truth for the tests, should I:
1. try to remove the os dependence in the tests
2. Only run these tests in unix systems
3. Remove the tests

**Comment by scovich** [comment]
> Since I'm using the filesystem to get a ground truth for the tests, should I:
> 
>     1. try to remove the os dependence in the tests
> 
>     2. Only run these tests in unix systems
> 
>     3. Remove the tests

It seems like we should either make the tests OS-independent, or have OS-specific code for each test so we verify it works? But then again this is only for local filesystem which is of little interest to prod code. The really useful test would validate that object store returns millisecond resolution timestamps, but that would require tests that access cloud storage.

**Comment by OussamaSaoudi-db** [comment]
> this is only for local filesystem which is of little interest to prod code. The really useful test would validate that object store returns millisecond resolution timestamps, but that would require tests that access cloud storage.
Does that mean I should remove the tests for now?

File: kernel/src/engine/sync/fs_client.rs [hunk 1]
```diff
@@ -51,19 +51,19 @@ impl FileSystemClient for SyncFilesystemClient {
...
-                                    Ok(d) => d.as_secs(),
-                                    Err(_) => 0,
-                                },
-                            )
+                            .ok()
+                            .and_then(|modified| {
+                                modified.duration_since(SystemTime::UNIX_EPOCH).ok()
+                            })
+                            .and_then(|modified| modified.as_millis().try_into().ok())
                             .unwrap_or(0);
```
Comment (scovich) [inline kernel/src/engine/sync/fs_client.rs:61]: Are we converting Result to Option three times here? Why is that needed? 
```suggestion
                            .and_then(|modified| modified.duration_since(SystemTime::UNIX_EPOCH))
                            .and_then(|modified| modified.as_millis().try_into())
                            .unwrap_or(0);
```
Or do types not match in some annoying way?

Comment (scovich) [reply kernel/src/engine/sync/fs_client.rs:61]: Honestly, all this monadic chaining feels like a poor substitute for a missing helper method...

```rust
impl TryFrom<DirEntry> for FileMeta {
    type Error = Error;

    fn try_from(ent: DirEntry) -> DeltaResult<FileMeta> {
        let metadata = ent.metadata()?;
        let last_modified = metadata.modified()?.duration_since(SystemTime::UNIX_EPOCH)?;
        Ok(FileMeta {
            location: Url::from_file_path(ent.path())?,
            last_modified: last_modified.as_millis().try_into()?,
            size: metadata.len() as usize,
        })
    }
}
```
and then
```rust

            let it = all_ents
                .into_iter()
                .sorted_by_key(|ent| ent.path())
                .map(TryFrom::try_from);
```

File: kernel/src/engine/sync/fs_client.rs [hunk 2]
```diff
@@ -97,20 +97,53 @@ impl FileSystemClient for SyncFilesystemClient {
...
+        let url = Url::from_file_path(url_path).unwrap();
+        let list: Vec<_> = client.list_from(&url)?.try_collect()?;
+
+        let metadata = f.metadata()?;
+        // We assert that the timestamp is in milliseconds
+        let expected_timestamp = metadata
+            .modified()?
+            .duration_since(UNIX_EPOCH)?
+            .as_millis()
+            .try_into()?;
```
Comment (scovich) [inline kernel/src/engine/sync/fs_client.rs:135]: This is the exact same code the iterator uses... so it's not actually testing for correctness but rather testing that the code is equivalent. Not sure what the better test would be? Maybe we write a file and manually set its mtime, then verify the FileMeta has the same timestamp? 

(same criticism applies to the default client test)

Comment (zachschuermann) [reply kernel/src/engine/sync/fs_client.rs:135]: maybe create the file then just assert that the timestamp we get back is within 10s or 60s or something?

File: kernel/src/engine/sync/fs_client.rs (hunk 1 above)
Comment (OussamaSaoudi-db) [reply kernel/src/engine/sync/fs_client.rs]: Woah that's clean 👌  I'm doing a lot of map_err, and I'm wondering if we should do `From` conversions for common things like `Url::from_*` and `TryInto` 🤔

File: kernel/src/engine/sync/fs_client.rs (hunk 2 above)
Comment (OussamaSaoudi-db) [reply kernel/src/engine/sync/fs_client.rs]: Implemented a test that checks that the filemeta is from the last minute.

File: kernel/src/engine/default/filesystem.rs [hunk 3]
```diff
@@ -216,6 +218,33 @@ mod tests {
         assert_eq!(data[2], Bytes::from("el-da"));
     }
 
+    #[tokio::test]
+    async fn test_file_meta_is_correct() {
+        let store = Arc::new(InMemory::new());
+
+        let begin_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        // The [`FileMeta`]s must be greater than 1 minute ago
+        let allowed_time = begin_time - Duration::from_secs(60);
```
Comment (scovich) [inline kernel/src/engine/default/filesystem.rs:227]: qq: Why do we need such a big safety margin, out of curiosity? All files are PUT after this timestamp, so they should not have smaller timestamps?

**Review by scovich — APPROVED** [review]
LGTM

File: kernel/src/engine/sync/fs_client.rs (hunk 1 above)
Comment (zachschuermann) [reply kernel/src/engine/sync/fs_client.rs]: yea seems reasonable? make an issue for us to think about?

File: kernel/src/engine/default/filesystem.rs (hunk 3 above)
Comment (zachschuermann) [reply kernel/src/engine/default/filesystem.rs:227]: Could we also just assert the file_meta time is < current time (and could include a little safety margin as well)? Right now we only have a bound on one end (I don't think this would catch the case of us accidentally returning nanos instead of millis for example)

File: kernel/src/engine/sync/fs_client.rs [hunk 4]
```diff
@@ -111,6 +92,30 @@ mod tests {
...
+        let mut f = File::create(path)?;
+        writeln!(f, "null")?;
+        f.flush()?;
+
+        let url_path = tmp_dir.path().join(get_json_filename(1));
+        let url = Url::from_file_path(url_path).unwrap();
+        let list: Vec<_> = client.list_from(&url)?.try_collect()?;
+        for meta in list.iter() {
+            let meta_time = Duration::from_millis(meta.last_modified.try_into()?);
+            assert!(allowed_time < meta_time,);
```
Comment (zachschuermann) [inline kernel/src/engine/sync/fs_client.rs:114]:
```suggestion
            assert!(allowed_time < meta_time);
```

**Review by zachschuermann — APPROVED** [review]
LGTM pending one fix!

File: kernel/src/engine/sync/fs_client.rs (hunk 1 above)
Comment (scovich) [reply kernel/src/engine/sync/fs_client.rs]: We already _have_ the `From` conversion for `url::ParseError`, so `?` should Just Work? The problem with `TryInto` is the error type is parametrized, so there's no way to know in advance what might be needed. We'd have to go case by case (when in doubt, try `?` and hopefully there's already an `impl From` for it to use).
